crates/red_knot_python_semantic/resources/mdtest/narrow/type.md

```diff
@@ -0,0 +1,127 @@
+# Narrowing for checks involving `type(x)`
+
+## `type(x) is C`
+
+```py
+class A: ...
+class B: ...
+
+def get_a_or_b() -> A | B: ...
+
+x = get_a_or_b()
+
+if type(x) is A:
+    reveal_type(x)  # revealed: A
+else:
+    # It would be wrong to infer `B` here. The type
+    # of `x` could be a subclass of `A`, so we need
+    # to infer the full union type:
+    reveal_type(x)  # revealed: A | B
+```
+
+## `type(x) is not C`
+
+```py
+class A: ...
+class B: ...
+
+def get_a_or_b() -> A | B: ...
+
+x = get_a_or_b()
+
+if type(x) is not A:
+    # Same reasoning as above: no narrowing should occur here.
+    reveal_type(x)  # revealed: A | B
+else:
+    reveal_type(x)  # revealed: A
+```
+
+## `type(x) == C`, `type(x) != C`
+
+No narrowing can occur for equality comparisons, since there might be a custom `__eq__`
+implementation on the metaclass:
```

Comment:
In fact, I think narrowing would also be safe if both classes had `@final` metaclasses, since the metaclass of a subclass of `A` must either be exactly equal to the metaclass of `A` or must be a proper subclass of the metaclass of `A`. E.g. narrowing would also be safe, I think, for something like this:

```py
from typing import final

@final
class Meta(type): ...

class A(metaclass=Meta): ...
class B(metaclass=Meta): ...

def get_a_or_b() -> A | B:
    return A()

x = get_a_or_b()

if type(x) != A:
    reveal_type(x)
```

Reply:
Thanks. Added a TODO comment.